File: drafting_service/orchestrator.py
"""AI Message Drafting Pipeline Orchestrator."""

from __future__ import annotations

import time
import uuid
from dataclasses import asdict
from typing import Any, Dict, List, Optional, TypedDict

from config import Config
from core.edit_logger import log_recruiter_edit
from core.leads import Lead, check_channel_eligibility, from_record
from core.rate_card import RateCardService
from claude_client import ClaudeClient
from draft_generator import Draft, generate_email, generate_linkedin
from evaluator import Evaluation, evaluate
from logger import get_logger

log = get_logger(__name__)


class PipelineDraftResult(TypedDict):
    draft_id: str
    channel: str
    lead_name: str
    subject: Optional[str]
    body: str
    verdict: str  # "SEND" | "HOLD" | "INELIGIBLE"
    evaluation: Dict[str, Any]
    flags: List[str]
    rate_applied: Optional[Dict[str, Any]]
    telemetry: Dict[str, Any]
    manual_override: bool


class DraftingOrchestrator:
    """6-Stage AI Message Drafting Pipeline Orchestrator."""

    def __init__(self, config: Config, rate_card: Optional[List[Dict[str, Any]]] = None):
        self.config = config
        self.client = ClaudeClient(config) if config.api_key else None
        self.rate_card_service = RateCardService(rate_card)

    def process_draft(
        self,
        lead_record: Dict[str, Any],
        channel: str = "email",
        manual_override: bool = False,
    ) -> PipelineDraftResult:
        start_time = time.monotonic()
        draft_id = f"draft_{uuid.uuid4().hex[:8]}"

        # Stage 1: Lead Ingestion
        lead = from_record(lead_record)

        # Stage 1.5: Automatic Eligibility Gate (bypassed entirely if manual_override)
        eligibility = check_channel_eligibility(lead, channel, manual_override=manual_override)
        if not eligibility.eligible:
            elapsed_ms = int((time.monotonic() - start_time) * 1000)
            log.info(
                "Drafting [%s] INELIGIBLE channel=%s lead=%s reason=%s",
                draft_id, channel, lead.first_name, eligibility.reason,
            )
            return {
                "draft_id": draft_id,
                "channel": channel,
                "lead_name": lead.first_name,
                "subject": None,
                "body": "",
                "verdict": "INELIGIBLE",
                "evaluation": {
                    "channel": channel,
                    "lead_name": lead.first_name,
                    "checks": [],
                    "programmatic_pass": False,
                    "send": False,
                    "flags": [eligibility.reason],
                },
                "flags": [eligibility.reason],
                "rate_applied": None,
                "telemetry": {
                    "model": None,
                    "latency_ms": 0,
                    "prompt_tokens": None,
                    "completion_tokens": None,
                    "total_execution_time_ms": elapsed_ms,
                },
                "manual_override": manual_override,
            }

        # Stage 2: Rate Card Lookup
        rate_match, rate_flag = self.rate_card_service.lookup_rate(
            source_lang=lead.source_language,
            target_lang=lead.target_language,
            service=lead.services[0] if lead.services else "Translation",
        )

        log.info("Drafting [%s] channel=%s lead=%s rate_match=%s", draft_id, channel, lead.first_name, bool(rate_match))

        # Stage 3: Prompt Building & LLM Generation, grounded strictly in
        # lead.grounding_facts() -- see prompts/prompt_builder.py's STRICT RULES.
        if not self.client:
            raise RuntimeError("CLAUDE_API_KEY is not configured in drafting service.")

        if channel == "email":
            draft = generate_email(self.client, self.config, lead, rate_match=rate_match, rate_flag=rate_flag)
        else:
            draft = generate_linkedin(self.client, self.config, lead, rate_match=rate_match, rate_flag=rate_flag)

        # Stage 4: Programmatic Rule Evaluation (evaluator.py)
        ev: Evaluation = evaluate(draft)

        # Stage 5: Approval Queue Formatting & Verdict Assignment
        verdict = "SEND" if ev.send else "HOLD"
        elapsed_ms = int((time.monotonic() - start_time) * 1000)

        telemetry = {
            "model": draft.model,
            "latency_ms": draft.latency_ms,
            "prompt_tokens": draft.prompt_tokens,
            "completion_tokens": draft.completion_tokens,
            "total_execution_time_ms": elapsed_ms,
        }

        log.info("Drafting [%s] complete -> verdict=%s (latency=%dms)", draft_id, verdict, elapsed_ms)

        return {
            "draft_id": draft_id,
            "channel": channel,
            "lead_name": lead.first_name,
            "subject": draft.subject,
            "body": draft.body,
            "verdict": verdict,
            "evaluation": ev.to_dict(),
            "flags": ev.flags,
            "rate_applied": rate_match,
            "telemetry": telemetry,
            "manual_override": manual_override,
        }

    def record_edit(self, draft_id: str, original_body: str, edited_body: str) -> Dict[str, Any]:
        """Stage 6: Record recruiter edits and compute similarity / edit-rate telemetry."""
        return log_recruiter_edit(draft_id, original_body, edited_body)

File: server/src/drafting/orchestrator.ts
/** AI Message Drafting Pipeline Orchestrator — direct port of
 * drafting_service/orchestrator.py, called in-process from the two
 * generate-draft route handlers instead of over HTTP. */

import { randomBytes } from "crypto";
import type { DraftingConfig } from "./config";
import { ClaudeClient } from "./claudeClient";
import { Draft, generateEmail, generateLinkedin } from "./draftGenerator";
import { EditLogResult, logRecruiterEdit } from "./editLogger";
import { checkChannelEligibility, fromRecord } from "./leads";
import { Evaluation, evaluate } from "./evaluator";
import { RateCardRow, RateCardService } from "./rateCard";

export interface PipelineDraftResult {
  draft_id: string;
  channel: string;
  lead_name: string;
  subject: string | null;
  body: string;
  verdict: "SEND" | "HOLD" | "INELIGIBLE";
  evaluation: Evaluation;
  flags: string[];
  rate_applied: RateCardRow | null;
  telemetry: {
    model: string | null;
    latency_ms: number;
    prompt_tokens: number | null;
    completion_tokens: number | null;
    total_execution_time_ms: number;
  };
  manual_override: boolean;
}

/** 6-Stage AI Message Drafting Pipeline Orchestrator. */
export class DraftingOrchestrator {
  private config: DraftingConfig;
  private client: ClaudeClient | null;
  private rateCardService: RateCardService;

  constructor(config: DraftingConfig, rateCard?: RateCardRow[] | null) {
    this.config = config;
    this.client = config.apiKey ? new ClaudeClient(config) : null;
    this.rateCardService = new RateCardService(rateCard);
  }

  async processDraft(
    leadRecord: Record<string, any>,
    channel: string = "email",
    manualOverride = false,
    // Our own database id for this lead -- kept as a separate top-level
    // parameter (like `channel`) rather than added to leadRecord/the Lead
    // class, which are a direct port of drafting_service's Python types and
    // never carried our id. Required (not optional) so a caller can't
    // silently omit it and fall back to the old static, unpersonalized
    // apply_url -- see ensureLinks() in draftGenerator.ts.
    leadId: string
  ): Promise<PipelineDraftResult> {
    const startTime = Date.now();
    const draftId = `draft_${randomBytes(4).toString("hex")}`;

    // Stage 1: Lead Ingestion
    const lead = fromRecord(leadRecord);

    // Stage 1.5: Automatic Eligibility Gate (bypassed entirely if manualOverride)
    const eligibility = checkChannelEligibility(lead, channel, manualOverride);
    if (!eligibility.eligible) {
      const elapsedMs = Date.now() - startTime;
      return {
        draft_id: draftId,
        channel,
        lead_name: lead.firstName,
        subject: null,
        body: "",
        verdict: "INELIGIBLE",
        evaluation: {
          channel,
          lead_name: lead.firstName,
          checks: [],
          programmatic_pass: false,
          send: false,
          flags: [eligibility.reason],
        },
        flags: [eligibility.reason],
        rate_applied: null,
        telemetry: {
          model: null,
          latency_ms: 0,
          prompt_tokens: null,
          completion_tokens: null,
          total_execution_time_ms: elapsedMs,
        },
        manual_override: manualOverride,
      };
    }

    // Stage 2: Rate Card Lookup
    const [rateMatch, rateFlag] = this.rateCardService.lookupRate(
      lead.sourceLanguage,
      lead.targetLanguage,
      lead.services.length ? lead.services[0] : "Translation"
    );

    // Stage 3: Prompt Building & LLM Generation, grounded strictly in
    // lead.groundingFacts() -- see promptBuilder.ts's VOICE_RULES.
    if (!this.client) {
      throw new Error("CLAUDE_API_KEY is not configured.");
    }

    const draft: Draft =
      channel === "email"
        ? await generateEmail(this.client, this.config, lead, leadId, rateMatch, rateFlag)
        : await generateLinkedin(this.client, this.config, lead, leadId, rateMatch, rateFlag);

    // Stage 4: Programmatic Rule Evaluation (evaluator.ts)
    const ev = evaluate(draft);

    // Stage 5: Approval Queue Formatting & Verdict Assignment
    const verdict: "SEND" | "HOLD" = ev.send ? "SEND" : "HOLD";
    const elapsedMs = Date.now() - startTime;

    const telemetry = {
      model: draft.model,
      latency_ms: draft.latency_ms,
      prompt_tokens: draft.prompt_tokens,
      completion_tokens: draft.completion_tokens,
      total_execution_time_ms: elapsedMs,
    };

    return {
      draft_id: draftId,
      channel,
      lead_name: lead.firstName,
      subject: draft.subject,
      body: draft.body,
      verdict,
      evaluation: ev,
      flags: ev.flags,
      rate_applied: rateMatch,
      telemetry,
      manual_override: manualOverride,
    };
  }

  /** Stage 6: Record recruiter edits and compute similarity / edit-rate telemetry. */
  recordEdit(draftId: string, originalBody: string, editedBody: string): EditLogResult {
    return logRecruiterEdit(draftId, originalBody, editedBody);
  }
}
